optuna/storages/_rdb/alembic/versions/v3.2.0.a_.py

```diff
@@ -0,0 +1,23 @@
+"""Add index to study_id column in trials table
+
+Revision ID: v3.2.0.a
+Revises: v3.0.0.d
+Create Date: 2023-02-25 13:21:00.730272
+
+"""
+from alembic import op
+
+
+# revision identifiers, used by Alembic.
+revision = "v3.2.0.a"
+down_revision = "v3.0.0.d"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.create_index(op.f("trials_study_id_key"), "trials", ["study_id"], unique=False)
+
+
+def downgrade():
+    op.drop_index(op.f("trials_study_id_key"), table_name="trials")
```

Comment:
Could you leave the following comment to let other developers know that this operation does not work with MySQL? Or you can simply skip the following operation when the sqlalchemy's dialect is mysql.

```suggestion
def downgrade():
    # Please note that the following operation does not work on MySQL due to a foreign key constraint.
    #
    # mysql> DROP INDEX ix_trials_study_id ON trials;
    # ERROR 1553 (HY000): Cannot drop index 'ix_trials_study_id': needed in a foreign key constraint
    op.drop_index(op.f("trials_study_id_key"), table_name="trials")
```

Reply:
[FYI] Here is the error log when executing this `downgrade()` operation on MySQL.

```
$ cd optuna/storages/_rdb
$ git diff
diff --git a/optuna/storages/_rdb/alembic.ini b/optuna/storages/_rdb/alembic.ini
index 52681f91a..9311cded0 100644
--- a/optuna/storages/_rdb/alembic.ini
+++ b/optuna/storages/_rdb/alembic.ini
@@ -36,7 +36,7 @@ script_location = alembic
 # output_encoding = utf-8

 # NOTE: This URL is only used when generating migration scripts.
-sqlalchemy.url = sqlite:///alembic.db
+sqlalchemy.url = mysql+pymysql://optuna:password@127.0.0.1:3306/optuna
```

```
$ alembic downgrade v3.0.0.d
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade v3.2.0.a -> v3.0.0.d, Add index to study_id column in trials table
Traceback (most recent call last):
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.OperationalError: (1091, "Can't DROP 'trials_study_id_key'; check that column/key exists")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/config.py", line 588, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/config.py", line 582, in main
    self.run_cmd(cfg, options)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/config.py", line 559, in run_cmd
    fn(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/command.py", line 366, in downgrade
    script.run_env()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/script/base.py", line 563, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 92, in load_python_file
    module = load_module_py(module_id, path)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 108, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "alembic/env.py", line 79, in <module>
    run_migrations_online()
  File "alembic/env.py", line 73, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/runtime/environment.py", line 851, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/runtime/migration.py", line 620, in run_migrations
    step.migration_fn(**kw)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/optuna/storages/_rdb/alembic/versions/v3.2.0.a_.py", line 23, in downgrade
    op.drop_index(op.f("trials_study_id_key"), table_name="trials")
  File "<string>", line 8, in drop_index
  File "<string>", line 3, in drop_index
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/operations/ops.py", line 1077, in drop_index
    return operations.invoke(op)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/operations/base.py", line 392, in invoke
    return fn(self, operation)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/operations/toimpl.py", line 104, in drop_index
    operations.impl.drop_index(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/ddl/impl.py", line 391, in drop_index
    self._exec(schema.DropIndex(index))
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/alembic/ddl/impl.py", line 197, in _exec
    return conn.execute(construct, multiparams)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1380, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py", line 80, in _execute_on_connection
    return connection._execute_ddl(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1472, in _execute_ddl
    ret = self._execute_context(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1943, in _execute_context
    self._handle_dbapi_exception(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2124, in _handle_dbapi_exception
    util.raise_(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/Users/c-bata/go/src/github.com/optuna/optuna/venv/lib/python3.9/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1091, "Can't DROP 'trials_study_id_key'; check that column/key exists")
[SQL:
DROP INDEX trials_study_id_key ON trials]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
```